### PR TITLE
feat: added processor to retrieve company name

### PR DESCRIPTION
**Pull request by franciscog-kainos**
### JIRA link
https://companieshouse.atlassian.net/browse/LFA-1814

### Change description
Added a FormActionProcessor that retrieves the company name from a company number.

### Work checklist

- [x] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/processors/CompanyNameProcessor.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        const token: string | undefined = session.get<ISignInInfo>(SessionKey.SignInInfo)?.access_token?.access_token;
+
+        if (!token) {
+            throw TOKEN_MISSING_ERROR;
+        }
+
+        const applicationData: ApplicationData | undefined = session.getExtraData(APPLICATION_DATA_KEY);
+        const companyNumber: string | undefined = applicationData?.appeal.penaltyIdentifier.companyNumber;
```
Comment (n-r-b) [inline src/controllers/processors/CompanyNameProcessor.ts:39]: Is the plan to validate the company penalty details, store them in the session and then get the company name? Currently from my understanding we process the form data and then finally persist the session

File: src/controllers/processors/CompanyNameProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { FormActionProcessor } from './FormActionProcessor';
+
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { CompaniesHouseSDK, OAuth2 } from 'app/modules/Types';
+
+
+export const SESSION_NOT_FOUND_ERROR: Error = new Error('Session Expected but was undefined');
+export const TOKEN_MISSING_ERROR: Error = new Error('Token was expected');
+export const COMPANY_NUMBER_UNDEFINED_ERROR: Error = new Error('Company number expected but was undefined');
+export const COMPANY_NUMBER_RETRIEVAL_ERROR = (companyNumber: string) => Error(`Could not retrieve company name for ${companyNumber}`);
```
Comment (n-r-b) [inline src/controllers/processors/CompanyNameProcessor.ts:17]: `COMPANY_NAME_RETRIEVAL_ERROR`?

File: src/controllers/processors/CompanyNameProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        if (!companyNumber) {
+            throw COMPANY_NUMBER_UNDEFINED_ERROR;
+        }
+
+        const profile: Resource<CompanyProfile> = await this
+            .chSdk(new OAuth2(token))
+            .companyProfile.getCompanyProfile(companyNumber);
+
+        const companyName = profile.resource?.companyName;
```
Comment (n-r-b) [inline src/controllers/processors/CompanyNameProcessor.ts:49]: type here would be nice :-)

File: src/controllers/processors/CompanyNameProcessor.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+        if (profile.httpStatusCode !== OK || !companyName) {
+            throw COMPANY_NUMBER_RETRIEVAL_ERROR(companyNumber);
+        }
+
+        applicationData!.appeal.penaltyIdentifier.companyName = companyName;
```
Comment (n-r-b) [inline src/controllers/processors/CompanyNameProcessor.ts:55]: if `applicationData` is undefined, would it be sensible to set it to `{}`?

File: test/controllers/processors/CompanyNameProcessor.test.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
+import 'reflect-metadata';
+
+import Substitute, { Arg } from '@fluffy-spoon/substitute';
+import ApiClient from 'ch-sdk-node/dist/client';
+import CompanyProfileService from 'ch-sdk-node/dist/services/company-profile/service';
+import { CompanyProfile } from 'ch-sdk-node/dist/services/company-profile/types';
+import { assert, expect } from 'chai';
+import { Request } from 'express';
+import { createSubstituteOf } from '../../SubstituteFactory';
```
Comment (n-r-b) [inline test/controllers/processors/CompanyNameProcessor.test.ts:9]: can you change these imported to the shortened version, i.e. `'test/SubstituteFactory'`

File: src/controllers/processors/CompanyNameProcessor.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            .chSdk(new OAuth2(token))
+            .companyProfile.getCompanyProfile(companyNumber);
+
+        const companyName: string | undefined = profile.resource?.companyName;
+
+        if (profile.httpStatusCode !== OK || !companyName) {
+            throw COMPANY_NAME_RETRIEVAL_ERROR(companyNumber);
+        }
+
+        request.body.companyName = companyName;
```
Comment (n-r-b) [inline src/controllers/processors/CompanyNameProcessor.ts:56]: when do we plan to update the session with the company name?

File: src/controllers/processors/CompanyNameProcessor.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import { inject } from 'inversify';
+import { provide } from 'inversify-binding-decorators';
+import { FormActionProcessor } from './FormActionProcessor';
+
+import { PenaltyIdentifier } from 'app/models/PenaltyIdentifier';
+import { CompaniesHouseSDK, OAuth2 } from 'app/modules/Types';
+
+
+export const SESSION_NOT_FOUND_ERROR: Error = new Error('Session Expected but was undefined');
+export const TOKEN_MISSING_ERROR: Error = new Error('Token was expected');
```
Comment (n-r-b) [inline src/controllers/processors/CompanyNameProcessor.ts:16]: Can we expand on this error message, for example 'Access token missing from session'

File: src/controllers/processors/CompanyNameProcessor.ts (hunk 6 above)
Comment (franciscog-kainos) [reply src/controllers/processors/CompanyNameProcessor.ts:56]: image

Comment (franciscog-kainos) [reply src/controllers/processors/CompanyNameProcessor.ts:56]: This is how the base controller behaves after running the processors. As seen here, it gets automatically saved in the session

File: test/controllers/processors/CompanyNameProcessor.test.ts [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        try {
+            await companyNameProcessor.process(request);
+            assert.fail('Expected to throw error');
+        } catch (err) {
+            companyProfileService.received().getCompanyProfile(companyNumber);
+            expect(err.message).to.equal(COMPANY_NAME_RETRIEVAL_ERROR(companyNumber).message);
+        }
+    });
+    it('should add the company name to the penaltyIdentifier in the session', async () => {
```
Comment (n-r-b) [inline test/controllers/processors/CompanyNameProcessor.test.ts:104]: how are we verifying the company name is added to the session here?

Comment (franciscog-kainos) [reply test/controllers/processors/CompanyNameProcessor.test.ts:104]: It's implicitly tested by BaseController tests. I'm only testing the functionality of this code.

File: src/controllers/processors/CompanyNameProcessor.ts (hunk 6 above)
Comment (n-r-b) [reply src/controllers/processors/CompanyNameProcessor.ts:56]: oh yes 👍 we can now remove the optional undefined operator on the `companyName` in Appeal?

Comment (franciscog-kainos) [reply src/controllers/processors/CompanyNameProcessor.ts:56]: No, unless we assign it a default value of '' here
image

Comment (n-r-b) [reply src/controllers/processors/CompanyNameProcessor.ts:56]: forgot about that! ok, doesn't hurt as it is then

File: test/controllers/processors/CompanyNameProcessor.test.ts (hunk 8 above)
Comment (n-r-b) [reply test/controllers/processors/CompanyNameProcessor.test.ts:104]: could just rename the test to reflect successful retrieval of company name or add a check to confirm the data has been added to body

**Pull request merged**